### PR TITLE
HTTP Reverse Proxy Transport

### PR DESCRIPTION
**Pull request by skidder**
HTTP reverse-proxy transport meant to address go-kit/kit#186

### CODE REVIEW

File: transport/httprp/server.go [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+	"golang.org/x/net/context"
+)
+
+// RequestFunc may take information from an HTTP request and put it into a
+// request context. BeforeFuncs are executed prior to invoking the
+// endpoint.
+type RequestFunc func(context.Context, *http.Request) context.Context
+
+// Server wraps an endpoint and implements http.Handler.
```
Comment (peterbourgon) [inline transport/httprp/server.go]: I guess it's not wrapping an endpoint anymore, but rather a proxying http.Handler, right?

File: transport/httprp/server.go [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+type ServerOption func(*Server)
+
+// ServerBefore functions are executed on the HTTP request object before the
+// request is decoded.
+func ServerBefore(before ...RequestFunc) ServerOption {
+	return func(s *Server) { s.before = before }
+}
+
+// ServeHTTP implements http.Handler.
+func (s Server) ServeHTTP(w http.ResponseWriter, r *http.Request) {
```
Comment (peterbourgon) [inline transport/httprp/server.go]: At first I thought this was a bug, but yeah, you _can_ define this on a value receiver.. neat!

File: transport/httprp/server_test.go [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+			w.Write([]byte("hey"))
+		}))
+	defer originServer.Close()
+	originURL, _ := url.Parse(originServer.URL)
+
+	handler := httptransport.NewServer(
+		context.Background(),
+		originURL,
+		httptransport.ServerBefore(func(ctx context.Context, r *http.Request) context.Context {
+			r.Header.Add("X-TEST-HEADER", "go-kit-proxy")
```
Comment (peterbourgon) [inline transport/httprp/server_test.go]: Would you mind extracting these strings to constants in the test function scope?

```go
const (
    headerKey = "X-TEST-HEADER"
    headerVal = "go-kit-proxy"
)
```

**Comment by peterbourgon** [comment]
Thank you for the very elegant and straightforward contribution! Just a few small things to fix up, and then I'll merge straight away.

**Comment by skidder** [comment]
Thanks again for the feedback Peter, I've incorporated your suggestions into the feature branch.

**Comment by peterbourgon** [comment]
Awesome! Thank you for the contribution!

**Pull request merged**